pkg/storaged/client.js

```diff
@@ -206,6 +214,200 @@
+        client.uuids_btrfs_default_subvol = btrfs_default_subvol;
+        debug("btrfs_pol default subvolumes:", btrfs_default_subvol);
+        client.update();
+    }
+}
+
+function btrfs_findmnt_poll() {
+    if (!client.btrfs_mounts)
+        client.btrfs_mounts = { };
+
+    const update_btrfs_mounts = output => {
+        const btrfs_mounts = {};
+        try {
+            // Extract the data into a { uuid: { subvolid: { subvol, target } } }
+            const mounts = JSON.parse(output);
+            if ("filesystems" in mounts) {
+                for (const fs of mounts.filesystems) {
+                    const subvolid_match = fs.options.match(/subvolid=(?<subvolid>\d+)/);
+                    const subvol_match = fs.options.match(/subvol=(?<subvol>[\w\\/]+)/);
+                    if (!subvolid_match && !subvol_match) {
+                        console.warn("findmnt entry without subvol and subvolid", fs);
+                        break;
+                    }
+
+                    const { subvolid } = subvolid_match.groups;
+                    const { subvol } = subvol_match.groups;
+                    const subvolume = {
+                        pathname: subvol,
+                        id: subvolid,
+                        mount_points: [fs.target],
+                    };
+
+                    if (!(fs.uuid in btrfs_mounts)) {
+                        btrfs_mounts[fs.uuid] = { };
+                    }
+
+                    // We need to handle multiple mounts, they are listed seperate.
+                    if (subvolid in btrfs_mounts[fs.uuid]) {
+                        btrfs_mounts[fs.uuid][subvolid].mount_points.push(fs.target);
+                    } else {
+                        btrfs_mounts[fs.uuid][subvolid] = subvolume;
+                    }
+                }
+            }
+        } catch (exc) {
+            if (exc.message)
+                console.error("unable to parse findmnt JSON output", exc);
+        }
+
+        // Update client state
+        if (!deep_equal(client.btrfs_mounts, btrfs_mounts)) {
+            client.btrfs_mounts = btrfs_mounts;
+            debug("btrfs_findmnt_poll mounts:", client.btrfs_mounts);
+            client.update();
+        }
+    };
+
+    const findmnt_poll = () => {
+        return cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "--poll"], { superuser: true, err: "message" }).stream(() => {
+            cockpit.spawn(["findmnt", "--type", "btrfs", "--mtab", "-o", "UUID,OPTIONS,TARGET", "--json"],
+                          { superuser: true, err: "message" }).then(output => update_btrfs_mounts(output)).catch(err => {
+                // When there are no btrfs filesystems left this can fail and thus we need to manually reset the mount info.
+                client.btrfs_mounts = {};
+                client.update();
+                if (err.message) {
+                    console.error("findmnt exited with an error", err);
```

Comment:
These 4 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19690-20231215-090815-3e4338b7-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)